//
// This file was generated by the Eclipse Implementation of JAXB, v4.0.5 
// See https://eclipse-ee4j.github.io/jaxb-ri 
// Any modifications to this file will be lost upon recompilation of the source schema. 
//


package org.uniprot;

import java.util.ArrayList;
import java.util.List;
import jakarta.xml.bind.annotation.XmlAccessType;
import jakarta.xml.bind.annotation.XmlAccessorType;
import jakarta.xml.bind.annotation.XmlAttribute;
import jakarta.xml.bind.annotation.XmlElement;
import jakarta.xml.bind.annotation.XmlType;


/**
 * Describes different types of sequence annotations.
 *             Equivalent to the flat file FT-line.
 * 
 * <p>Java class for featureType complex type</p>.
 * 
 * <p>The following schema fragment specifies the expected content contained within this class.</p>
 * 
 * <pre>{@code
 * <complexType name="featureType">
 *   <complexContent>
 *     <restriction base="{http://www.w3.org/2001/XMLSchema}anyType">
 *       <sequence>
 *         <element name="original" type="{http://www.w3.org/2001/XMLSchema}string" minOccurs="0"/>
 *         <element name="variation" type="{http://www.w3.org/2001/XMLSchema}string" maxOccurs="unbounded" minOccurs="0"/>
 *         <element name="location" type="{http://uniprot.org/uniprot}locationType"/>
 *         <element name="ligand" type="{http://uniprot.org/uniprot}ligandType" minOccurs="0"/>
 *         <element name="ligandPart" type="{http://uniprot.org/uniprot}ligandPartType" minOccurs="0"/>
 *       </sequence>
 *       <attribute name="type" use="required">
 *         <simpleType>
 *           <restriction base="{http://www.w3.org/2001/XMLSchema}string">
 *             <enumeration value="active site"/>
 *             <enumeration value="binding site"/>
 *             <enumeration value="chain"/>
 *             <enumeration value="coiled-coil region"/>
 *             <enumeration value="compositionally biased region"/>
 *             <enumeration value="cross-link"/>
 *             <enumeration value="disulfide bond"/>
 *             <enumeration value="DNA-binding region"/>
 *             <enumeration value="domain"/>
 *             <enumeration value="glycosylation site"/>
 *             <enumeration value="helix"/>
 *             <enumeration value="initiator methionine"/>
 *             <enumeration value="lipid moiety-binding region"/>
 *             <enumeration value="modified residue"/>
 *             <enumeration value="mutagenesis site"/>
 *             <enumeration value="non-consecutive residues"/>
 *             <enumeration value="non-terminal residue"/>
 *             <enumeration value="peptide"/>
 *             <enumeration value="propeptide"/>
 *             <enumeration value="region of interest"/>
 *             <enumeration value="repeat"/>
 *             <enumeration value="non-standard amino acid"/>
 *             <enumeration value="sequence conflict"/>
 *             <enumeration value="sequence variant"/>
 *             <enumeration value="short sequence motif"/>
 *             <enumeration value="signal peptide"/>
 *             <enumeration value="site"/>
 *             <enumeration value="splice variant"/>
 *             <enumeration value="strand"/>
 *             <enumeration value="topological domain"/>
 *             <enumeration value="transit peptide"/>
 *             <enumeration value="transmembrane region"/>
 *             <enumeration value="turn"/>
 *             <enumeration value="unsure residue"/>
 *             <enumeration value="zinc finger region"/>
 *             <enumeration value="intramembrane region"/>
 *           </restriction>
 *         </simpleType>
 *       </attribute>
 *       <attribute name="id" type="{http://www.w3.org/2001/XMLSchema}string" />
 *       <attribute name="description" type="{http://www.w3.org/2001/XMLSchema}string" />
 *       <attribute name="evidence" type="{http://uniprot.org/uniprot}intListType" />
 *       <attribute name="ref" type="{http://www.w3.org/2001/XMLSchema}string" />
 *     </restriction>
 *   </complexContent>
 * </complexType>
 * }</pre>
 * 
 * 
 */
@XmlAccessorType(XmlAccessType.FIELD)
@XmlType(name = "featureType", propOrder = {
    "original",
    "variation",
    "location",
    "ligand",
    "ligandPart"
})
public class FeatureType {

    /**
     * Describes the original sequence in annotations that describe natural or artifical sequence variations.
     * 
     */
    protected String original;
    /**
     * Describes the variant sequence in annotations that describe natural or artifical sequence variations.
     * 
     */
    protected List<String> variation;
    /**
     * Describes the sequence coordinates of the annotation.
     * 
     */
    @XmlElement(required = true)
    protected LocationType location;
    /**
     * Describes the chemical entity that is bound in annotations that describe binding sites.
     * 
     */
    protected LigandType ligand;
    /**
     * Describes the specific part of a molecule that is bound in annotations that describe binding sites.
     * 
     */
    protected LigandPartType ligandPart;
    /**
     * Describes the type of a sequence annotation.
     *                 Equivalent to the flat file FT feature keys, but using full terms instead of acronyms.
     * 
     */
    @XmlAttribute(name = "type", required = true)
    protected String type;
    @XmlAttribute(name = "id")
    protected String id;
    @XmlAttribute(name = "description")
    protected String description;
    @XmlAttribute(name = "evidence")
    protected List<Integer> evidence;
    @XmlAttribute(name = "ref")
    protected String ref;

    /**
     * Describes the original sequence in annotations that describe natural or artifical sequence variations.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getOriginal() {
        return original;
    }

    /**
     * Sets the value of the original property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     * @see #getOriginal()
     */
    public void setOriginal(String value) {
        this.original = value;
    }

    /**
     * Describes the variant sequence in annotations that describe natural or artifical sequence variations.
     * 
     * Gets the value of the variation property.
     * 
     * <p>This accessor method returns a reference to the live list,
     * not a snapshot. Therefore any modification you make to the
     * returned list will be present inside the JAXB object.
     * This is why there is not a <CODE>set</CODE> method for the variation property.</p>
     * 
     * <p>
     * For example, to add a new item, do as follows:
     * </p>
     * <pre>
     * getVariation().add(newItem);
     * </pre>
     * 
     * 
     * <p>
     * Objects of the following type(s) are allowed in the list
     * {@link String }
     * </p>
     * 
     * 
     * @return
     *     The value of the variation property.
     */
    public List<String> getVariation() {
        if (variation == null) {
            variation = new ArrayList<>();
        }
        return this.variation;
    }

    /**
     * Describes the sequence coordinates of the annotation.
     * 
     * @return
     *     possible object is
     *     {@link LocationType }
     *     
     */
    public LocationType getLocation() {
        return location;
    }

    /**
     * Sets the value of the location property.
     * 
     * @param value
     *     allowed object is
     *     {@link LocationType }
     *     
     * @see #getLocation()
     */
    public void setLocation(LocationType value) {
        this.location = value;
    }

    /**
     * Describes the chemical entity that is bound in annotations that describe binding sites.
     * 
     * @return
     *     possible object is
     *     {@link LigandType }
     *     
     */
    public LigandType getLigand() {
        return ligand;
    }

    /**
     * Sets the value of the ligand property.
     * 
     * @param value
     *     allowed object is
     *     {@link LigandType }
     *     
     * @see #getLigand()
     */
    public void setLigand(LigandType value) {
        this.ligand = value;
    }

    /**
     * Describes the specific part of a molecule that is bound in annotations that describe binding sites.
     * 
     * @return
     *     possible object is
     *     {@link LigandPartType }
     *     
     */
    public LigandPartType getLigandPart() {
        return ligandPart;
    }

    /**
     * Sets the value of the ligandPart property.
     * 
     * @param value
     *     allowed object is
     *     {@link LigandPartType }
     *     
     * @see #getLigandPart()
     */
    public void setLigandPart(LigandPartType value) {
        this.ligandPart = value;
    }

    /**
     * Describes the type of a sequence annotation.
     *                 Equivalent to the flat file FT feature keys, but using full terms instead of acronyms.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getType() {
        return type;
    }

    /**
     * Sets the value of the type property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     * @see #getType()
     */
    public void setType(String value) {
        this.type = value;
    }

    /**
     * Gets the value of the id property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getId() {
        return id;
    }

    /**
     * Sets the value of the id property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setId(String value) {
        this.id = value;
    }

    /**
     * Gets the value of the description property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getDescription() {
        return description;
    }

    /**
     * Sets the value of the description property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setDescription(String value) {
        this.description = value;
    }

    /**
     * Gets the value of the evidence property.
     * 
     * <p>This accessor method returns a reference to the live list,
     * not a snapshot. Therefore any modification you make to the
     * returned list will be present inside the JAXB object.
     * This is why there is not a <CODE>set</CODE> method for the evidence property.</p>
     * 
     * <p>
     * For example, to add a new item, do as follows:
     * </p>
     * <pre>
     * getEvidence().add(newItem);
     * </pre>
     * 
     * 
     * <p>
     * Objects of the following type(s) are allowed in the list
     * {@link Integer }
     * </p>
     * 
     * 
     * @return
     *     The value of the evidence property.
     */
    public List<Integer> getEvidence() {
        if (evidence == null) {
            evidence = new ArrayList<>();
        }
        return this.evidence;
    }

    /**
     * Gets the value of the ref property.
     * 
     * @return
     *     possible object is
     *     {@link String }
     *     
     */
    public String getRef() {
        return ref;
    }

    /**
     * Sets the value of the ref property.
     * 
     * @param value
     *     allowed object is
     *     {@link String }
     *     
     */
    public void setRef(String value) {
        this.ref = value;
    }

}
